### PR TITLE
Package diagnostics

### PR DESCRIPTION
**Pull request by JJJianJiao**
Jira Ticket: https://unity3d.atlassian.net/browse/GPA-39

### CODE REVIEW

File: Tests/Editor/PackagesTests.cs [hunk 1]
```diff
@@ -47,5 +70,31 @@ public void InstalledPackage_IsReported(string description, string name, string
...
+        }
+
+        [Test]
+        public void RecommandedPreviewPackage_IsReproted()
+        {
+            var issuePackages = Analyze(IssueCategory.PackageVersion);
+            var matchIssue = issuePackages.FirstOrDefault(issue => issue.customProperties[0] == "com.unity.services.vivox");
+
+            Assert.IsNotNull(matchIssue, "Cannot find the upgrade pacakge: com.unity.services.vivox");
+            Assert.AreEqual(matchIssue.customProperties[0], "com.unity.services.vivox");
```
Comment (mtrive) [inline Tests/Editor/PackagesTests.cs:94]: same here

File: Tests/Editor/PackagesTests.cs [hunk 2]
```diff
@@ -47,5 +70,31 @@ public void InstalledPackage_IsReported(string description, string name, string
...
             }
         }
+
+        [Test]
+        public void RecommandedUpgradePackage_IsReproted()
+        {
+            var issuePackages = Analyze(IssueCategory.PackageVersion);
+            var matchIssue = issuePackages.FirstOrDefault(issue => issue.customProperties[0] == "com.unity.2d.pixel-perfect");
+
+            Assert.IsNotNull(matchIssue, "Cannot find the upgrade pacakge: com.unity.2d.pixel-perfect");
```
Comment (mtrive) [inline Tests/Editor/PackagesTests.cs:80]: typo in pacakge

File: Tests/Editor/PackagesTests.cs [hunk 3]
```diff
@@ -47,5 +70,31 @@ public void InstalledPackage_IsReported(string description, string name, string
...
+            Assert.AreEqual(matchIssue.customProperties[0], "com.unity.2d.pixel-perfect");
+            Assert.AreEqual(matchIssue.customProperties[1], "3.0.2");
+            Assert.AreEqual(matchIssue.customProperties[2], "4.0.1");
```
Comment (mtrive) [inline Tests/Editor/PackagesTests.cs:83]: The recommended version will vary depending on the version of Unity. Perhaps we should just check that the recommended version is higher than 3.0.2 ?

File: Editor/Modules/PackagesModule.cs [hunk 4]
```diff
@@ -13,31 +10,71 @@ namespace Unity.ProjectAuditor.Editor.Modules
...
+                new PropertyDefinition { type = PropertyTypeUtil.FromCustom(PackageVersionProperty.PackageID), format = PropertyFormat.String, name = "ID", longName = "Package ID", defaultGroup = true},
+                new PropertyDefinition { type = PropertyTypeUtil.FromCustom(PackageVersionProperty.CurrentVersion), format = PropertyFormat.String, name = "Current Version" },
+                new PropertyDefinition { type = PropertyTypeUtil.FromCustom(PackageVersionProperty.RecommendedVersion), format = PropertyFormat.String, name = "Recommended Version"},
+                new PropertyDefinition { type = PropertyTypeUtil.FromCustom(PackageVersionProperty.Experimental), format = PropertyFormat.Bool, name = "Experimental/Preview" }
+            }
+        };
+
+
+        static readonly ProblemDescriptor k_RecommendPackageUpgrade  = new ProblemDescriptor(
+            "PKG0001",
```
Comment (mtrive) [inline Editor/Modules/PackagesModule.cs:58]: Let's change the IDs to use to _PAP_ (Project Auditor Package) as prefix

File: Tests/Editor/PackagesTests.cs [hunk 5]
```diff
@@ -2,11 +2,34 @@
...
+        public void SetUp()
+        {
+            AddRequest AddRequest = Client.Add("com.unity.2d.pixel-perfect@3.0.2");
+            while (AddRequest.Status != StatusCode.Success) {}
+            AddRequest = Client.Add("com.unity.services.vivox");
+            while (AddRequest.Status != StatusCode.Success) {}
+            //CompilationPipeline.RequestScriptCompilation();   //Use this function to make the Unity Editor recompile every Script in your Project asynchronously. When the compilation finishes, the Unity Editor then loads the compiled assemblies. https://docs.unity3d.com/2020.1/Documentation/ScriptReference/Compilation.CompilationPipeline.RequestScriptCompilation.html
+        }
+
+        [OneTimeTearDown]
```
Comment (mtrive) [inline Tests/Editor/PackagesTests.cs:23]: Can we remove this?

**Review by mtrive — APPROVED** [review]
Looks pretty good overall. Approved subject to fixing IDs and unit tests.

File: Tests/Editor/PackagesTests.cs (hunk 3 above)
Comment (JJJianJiao) [reply Tests/Editor/PackagesTests.cs:81]: Hi Marco, I do not get your point. 3.0.2 is the current version. 4.0.1 is the recommended version.

**Pull request merged**